addons/website_sale/controllers/main.py

```diff
@@ -249,10 +249,52 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
         Product = request.env['product.template'].with_context(bin_size=True)
 
         search_product = Product.search(domain, order=self._get_search_order(post))
+        product_count = len(search_product)
```

Comment:
This is the root of all evil in this controller.
If you look carefully, you don't actually *need* `search_product`, but instead:
* `len(search_product)`, which may be given by `search_count`...
* injecting `search_product.ids` into a domain (a subquery would do the job)
* `search_product[offset: offset + ppg]`, which may be searched in a specific way

What I suggest is to use method `_search()`, which returns a `Query` object that represents the product ids:
```python
# this Query object replaces the recordset 'search_product'
product_ids = Product._search(domain)
```
You can inject `product_ids` into domains: the domain converter (expression.py) should automatically inject a subquery in the main query, and the main query is left to be optimized by PostgreSQL:
```python
search_categories = Category.search([('product_tmpl_ids', 'in', product_ids)] + website_domain).parents_and_self
```
For the actual product search, simply make it explicit:
```python
# this replaces len(search_product)
product_count = Product.search_count(domain)

# this replaces search_product[offset: offset + ppg]
products = Product.search(domain, order=self._get_search_order(post), offset=offset, limit=ppg)
```

Reply:
I had this impression too, but the truth is that searching all records is faster than reading ``search_count`` + ``search(..., limit=..., offset=...)``. Something like 240 ms vs 190+190 ms